### PR TITLE
Objectstore global tools

### PR DESCRIPTION
**Pull request by SimonRichardson**
~~Requires https://github.com/juju/juju/pull/16525 to land first.~~

-----

The following pushes the objectstorage from the worker to the
tool agent storage requester. The tool's code is a bit overly complicated,
as we record the blob metadata in multiple locations. In the
toolsmetadata, managedstoredresources, blobstore collections. Most of
the information is duplicated, it might be wise when we convert the
tools state to the dqlite domain services to reduce the amount of
duplication, or at the very least use RI to remove duplication.

The code changes here, just push the objectstore to the binary storage
type so that it can use the underlying managed storage. I've removed
the layered storage, as we no longer want to have the option of
hierarchical storage for agent tools. Instead, just a global namespace
will suffice. In the new world, all the information will be in
the global controller namespace. For the state packages, we use the
controller model-uuid. This will simplify a lot of the code we have.

If we can chip away at this complexity as we go, I think it will have
a nice model of what we actually do, rather than an over-engineered
view of what we might have wanted at some point.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju enable-ha
$ juju upgrade-controller --build-agent
$ juju upgrade-model
```

## Links

**Jira card:** JUJU-4890

### CODE REVIEW

File: apiserver/facades/client/machinemanager/machinemanager.go [hunk 1]
```diff
@@ -356,7 +359,7 @@ func (mm *MachineManagerAPI) ProvisioningScript(ctx context.Context, args params
 		return result, errors.Trace(err)
 	}
 
-	icfg, err := InstanceConfig(ctx, mm.controllerConfigService, st, mm.st, mm.cloudService, mm.credentialService, args.MachineId, args.Nonce, args.DataDir)
+	icfg, err := InstanceConfig(ctx, mm.controllerConfigService, st, mm.st, mm.cloudService, mm.credentialService, mm.controllerStore, args.MachineId, args.Nonce, args.DataDir)
```
Comment (manadart) [inline apiserver/facades/client/machinemanager/machinemanager.go:362]: For cases like this, I would be tempted to create an encapsulating service, embed all the constituent services, and pass it as a single argument.

File: state/binarystorage/binarystorage.go [hunk 2]
```diff
@@ -20,9 +20,27 @@ import (
 
 var logger = loggo.GetLogger("juju.state.binarystorage")
 
+// ManagedStorage instances persist data for a bucket, for a user, or globally.
+// (Only bucket storage is currently implemented).
+type ManagedStorage interface {
+	// GetForBucket returns a reader for data at path, namespaced to the bucket.
```
Comment (manadart) [inline state/binarystorage/binarystorage.go:26]: Wrong comment - "Get".

File: state/binarystorage/binarystorage.go [hunk 3]
```diff
@@ -20,9 +20,27 @@ import (
...
+	// If the data is still being uploaded and is not fully written yet,
+	// an ErrUploadPending error is returned. This means the path is valid but the caller
+	// should try again to retrieve the data.
+	Get(ctx context.Context, path string) (r io.ReadCloser, length int64, err error)
+
+	// PutForBucket stores data from reader at path, namespaced to the bucket.
```
Comment (manadart) [inline state/binarystorage/binarystorage.go:32]: Same for the other two.

**Review by manadart — APPROVED** [review]
Looks good; couple of things to take a look at.

File: state/binarystorage/binarystorage.go (hunk 2 above)
Comment (SimonRichardson) [reply state/binarystorage/binarystorage.go:26]: This was fixed in later commits.

**Pull request merged**